### ISSUE TITLE
Introduction of Wide Resnet implementation & gated implementation for tornado prediction

### ISSUE THREAD

**Issue by Higgs32584** [mit-ll/tornet#11]
Hi MIT-LL team,

I recently completed a project building on your TorNet framework, which I’ve made available here: https://github.com/Higgs32584/tornet-Higgins. The project includes a wide Resnet implementation, that from my rigorous testing, with cross validation, outperforms the current model available. the generation scripts are available in the scripts/tornado_detection folder, if any of you are interested.

I'd love to hear your thoughts and see if there's any potential for collaboration or contribution back to the original repository. Thank you for making this great tool publicly available.

Just an FYI, both models have from_logits=False, and any custom functions should be present in the training script.

@markveillette
@tonyreisll

**Comment by markveillette** [comment]
LOVE THIS!   Thanks so much!  Can't wait to dig in and see what you did here.   I plan on creating a papers with code to track metrics, and this will be great addition.

**Comment by Higgs32584** [comment]
@markveillette of course! If you have any questions or are looking for any additional scripts don't hesitate to reach out.
I might have what you are looking for in the other branch, I'd just have to dig it up since the other branch is very messy.

Thank you!

**Comment by Higgs32584** [comment]
Hi! 

Just a heads up I made another breakthrough.

Feel free to check it out

**Comment by markveillette** [comment]
Wow,  really impressive!!  :)

The ensemble method with an AUCPR of 59% definitely beats our best model so far,  so that's very exciting!  

I'm  trying to find time to fully digest this great contribution.  I'm curious, have you tried visualizing the state of the model before the dense layers at the end?   One thing we like to do is create a "heatmap", similar to figures 11 and 12 in the paper.  This can be done on individual chips in TorNet,   similar to this image that is from my AMS talk this past year.

![Image](https://github.com/user-attachments/assets/e52805f1-7447-4bc6-a77a-60f658fdda2c)

In our case, we created the "heatmap" by chopping off layer at the end that reduces everything to a scalar output.  

We created the `grad` image by computing the gradient of the model w.r.t. the inputs.   The gist of the computation is as follows:

```python
# model is trained tf model
# model_x is dict of inputs
with tf.GradientTape() as tape:
    tape.watch(model_x)
    logits = model(model_x, training=False)

gradient = tape.gradient(logits, model_x)
```

It really shows in the image the how model reacts and often pinpoints the precise location of the tornado within the chip.  Might be interesting to create these images with your better models.

BTW, I hope to set up a Papers with code page to track the best performing models.  I'd like to create a score card like the one you're showing,  and include different views of the datasets (e.g. one excluding the warnings, as those can be tricky).

Thanks again for the great work.

**Comment by Higgs32584** [comment]
so heres an example I got up, I looked at my model and I realized that the attention map is likely your best bet. The final layers are dense, so you can't reconstruct them. Still, you can reconstruct attention, which is the deepest convolutional layer in the model


``` python
import matplotlib.pyplot as plt
import numpy as np
import tensorflow as tf
from tornet.data.constants import CHANNEL_MIN_MAX
from tornet.display.display import get_cmap
cnn = tf.keras.models.load_model("fold5seed99TH4242.keras",safe_mode=False,compile=False,custom_objects=custom_objects)
heatmap_model = tf.keras.Model(
    inputs=cnn.inputs,
    outputs=cnn.get_layer("attention_map").output
)
for x, y in ds_test:
    # Convert selected inputs to tensors
    x_tensor_dict = {
        k: tf.convert_to_tensor(x[k]) for k in [
            "DBZ", "VEL", "KDP", "RHOHV", "ZDR", "WIDTH", "range_folded_mask", "coordinates"
        ]
    }
    heatmap_output = heatmap_model(x_tensor_dict)[0, ..., 0].numpy()  # shape (H, W)
    heatmap = (heatmap_output - np.min(heatmap_output)) / (np.max(heatmap_output) - np.min(heatmap_output) + 1e-8)

    with tf.GradientTape() as tape:
        # Watch all input tensors
        for v in x_tensor_dict.values():
            tape.watch(v)

        # Forward pass
        logits = cnn(x_tensor_dict, training=False)
        prob = tf.squeeze(logits)  # scalar: P_tornado
    if(prob >= 0.4242):
        channel='DBZ'
        gradient = tape.gradient(prob, x_tensor_dict[channel])  # (1, H, W, C)
        gradient = gradient[0]  # remove batch dimension -> (H, W, C)
        grad_magnitude = tf.norm(gradient, axis=-1).numpy()  # (H, W)

        raw_image = x_tensor_dict[channel][0, ..., 0].numpy()
        vmin, vmax = CHANNEL_MIN_MAX[channel]
        normalized_image = (raw_image - vmin) / (vmax - vmin)
        # normalized_image = np.clip(normalized_image, 0.0, 1.0)  # Optional: ensure values stay in [0, 1]

        heatmap_output = heatmap_model(x_tensor_dict)[0, ..., 0].numpy()  # shape (H, W)
        heatmap = (heatmap_output - np.min(heatmap_output)) / (np.max(heatmap_output) - np.min(heatmap_output) + 1e-8)
        fig, axes = plt.subplots(1, 3, figsize=(14, 5))
        cmap,norm=get_cmap(channel)
        im0 = axes[0].imshow(normalized_image, cmap=cmap)
        axes[0].set_title(f"{channel}\n t = 05 min")
        axes[0].axis("off")
        fig.colorbar(im0, ax=axes[0], fraction=0.046, pad=0.04, label=f"{channel} (normalized)")

        # 2. Heatmap (model relevance or just probability)
        im1 = axes[1].imshow(heatmap, cmap='viridis')
        axes[1].set_title("attention map\n t = 05 min")
        axes[1].axis("off")
        fig.colorbar(im1, ax=axes[1], fraction=0.046, pad=0.04, label="Attention")

        # 3. Gradient magnitude heatmap
        vmax = np.max(np.abs(grad_magnitude))
        im2 = axes[2].imshow(grad_magnitude, cmap='viridis', vmin=0, vmax=vmax if vmax > 0 else 1e-4)
        axes[2].set_title("grad\n t = 05 min")
        axes[2].axis("off")
        fig.colorbar(im2, ax=axes[2], fraction=0.046, pad=0.04, label=r"$\|\nabla_{channel} P_{tornado}\|$")

        # Global title
        event_id = x.get("event_id", ["unknown"])[0]
        ef_score = x.get("ef_number", ["?"])[0]
        fig.suptitle(f"Event id: {event_id}, Category: Tornado, EF Score: {ef_score}, $P_{{tornado}}$ = {prob.numpy():.2f}", fontsize=14)

        plt.tight_layout()
        plt.show()
```




![Image](https://github.com/user-attachments/assets/c5f86883-4c6b-4708-9697-53a6e1cd0a19)

**Comment by Higgs32584** [comment]
@markveillette Hi mark, I also wanted to say one thing I am noticing about these models, or at least the successful ones. is that they seem to really fail to generalize the more and more parameters you add. 

Why this is I am not really sure, but it seems like you have to use very few parameters to force it to generalize better. Otherwise, it seems to latch onto unhelpful features.

**Comment by tonyreisll** [comment]
@Higgs32584 what you are describing is the bias-variance tradeoff. Increasing model complexity by adding more parameters will eventually lead to overfitting.

**Comment by Higgs32584** [comment]
@tonyreisll You’re right, though it’s interesting—typically, larger models perform at least as well as smaller ones, if not better. But in this case, increasing the number of parameters actually seems to hurt performance.

I keep thinking there’s a lower limit to how small the model can go without sacrificing predictive power, but I’ve gone from 230k to 102k to 60k parameters just last night by replacing the largest Conv2D layer with a depthwise separable block, and the performance is still nearly the same.

Interestingly, none of the starter models from here: https://keras.io/api/applications/ have come close to matching the performance of the smaller custom ones, even when using coordinates and a range_folded_mask. They tend to overfit on less useful features and struggle to get a val_AUCPR above 0.50. It almost feels like the model starts focusing on the right patterns only when its capacity is constrained.

**Comment by markveillette** [comment]
I definitely want to look at that attention map some more,  that seems really interesting and could be quite useful!

Regarding the above observations about overfitting, another factor to consider if that there may be mislabeled samples in TorNet, particularly in the warning class.  Not accounting for this in training could lead to the effect you describe, where bigger models overfit to these cases and ultimatley worsen the generalizability.   Label smoothing or regularization are good counters for this.

**Comment by Higgs32584** [comment]
I am also noticing that Dilation Definitely affects the role of prediction as well. Here are some scores for tornado prediction for models trained with different dilations on the intermediate layer after the WRN that I have noticed. But Dilation 1 and 2 is where there are two conv2d filters side by side, one with a dilation of 1 and one with a dilation of 2, and the model learns which to take. Dilation 2 and dilation 4 is just one conv2d with the respective Dilation.


Total Number of Tornadoes in Validation 2013-2014: 1505
Total Correct in dilation1and2:  855
Total Correct in dilation2: 844
Total Correct in dilation4  829
UNIQUELY Correct in dilation12:  35
UNIQUELY Correct in dilation2:  25
UNIQUELY Correct in dilation4:  25
CORRECT in either dilation1and2 or 4: 917
CORRECT in either dilation1and2 or 2: 917
CORRECT in either dilation2 or 4 907
Correctly Classified in Any 3: 942
Correctly Classified in All 3: 729

We can definitely see that dilation can play a role in the classification at least for some of them. 


These, for instance, these are some particular edge cases where predictions diverged trained on different dilations.


![Image](https://github.com/user-attachments/assets/a0e22006-f0c4-41cc-bc2d-3c939fec4d5b)


![Image](https://github.com/user-attachments/assets/1b4a3aeb-3e9e-45f3-b33f-4a929208b91e)



![Image](https://github.com/user-attachments/assets/b608c893-f48d-46c0-8630-639f2f0bded4)



![Image](https://github.com/user-attachments/assets/7c7b947c-b27f-457c-bd2e-2151902b0f32)

**Comment by Higgs32584** [comment]
Just updated this will probably be my last update! My schedule is getting busy! Thank you again for providing the dataset it was very interesting stuff!
